src/main/java/bio/terra/filesystem/FireStoreFileDao.java

```diff
@@ -0,0 +1,561 @@
+            if (FSObject.FSObjectType.fromLetter(dirObject.getObjectTypeLetter()) != FSObject.FSObjectType.DIRECTORY) {
+                throw new InvalidFileSystemObjectTypeException("Only a directory can be enumerated");
+            }
+
+            Query query =  firestore.collection(studyId.toString()).whereEqualTo("path", getFullPath(dirObject));
+            ApiFuture<QuerySnapshot> querySnapshot = xn.get(query);
+            List<QueryDocumentSnapshot> documents = querySnapshot.get().getDocuments();
+
+            List<FSObject> fsObjectList = new ArrayList<>();
+            for (QueryDocumentSnapshot document : documents) {
+                FireStoreObject fireStoreObject = document.toObject(FireStoreObject.class);
+                FSObject fsObject = makeFSObjectFromFileObject(fireStoreObject);
+                fsObjectList.add(fsObject);
+            }
+
+            return fsObjectList;
+        });
+
+        return fireStoreUtils.transactionGet("enumerate directory", transaction);
+    }
+
+    public FSObject retrieve(UUID studyId, UUID objectId) {
+        FSObject fsObject = retrieveByIdNoThrow(studyId, objectId);
+        if (fsObject == null) {
+            throw new FileSystemObjectNotFoundException("Object not found. Requested id is: " + objectId);
+        }
+        return fsObject;
+    }
+
+    public FSObject retrieveByIdNoThrow(UUID studyId, UUID objectId) {
+        ApiFuture<FSObject> transaction = firestore.runTransaction(xn -> {
+            QueryDocumentSnapshot docSnap = lookupByObjectId(studyId.toString(), objectId.toString(), xn);
+            if (docSnap == null) {
+                return null;
+            }
+            FireStoreObject currentObject = docSnap.toObject(FireStoreObject.class);
+            return makeFSObjectFromFileObject(currentObject);
+        });
+
+        return fireStoreUtils.transactionGet("retrieve by id", transaction);
+    }
+
+    public FSObject retrieveByPath(UUID studyId, String path) {
+        FSObject fsObject = retrieveByPathNoThrow(studyId, path);
+        if (fsObject == null) {
+            throw new FileSystemObjectNotFoundException("Object not found - path: '" + path + "'");
+        }
+        return fsObject;
+    }
+
+    public FSObject retrieveByPathNoThrow(UUID studyId, String path) {
+        DocumentReference docRef = firestore
+            .collection(studyId.toString())
+            .document(getDocumentName(path));
+
+        ApiFuture<DocumentSnapshot> docSnapFuture = docRef.get();
+        try {
+            DocumentSnapshot docSnap = docSnapFuture.get();
+            if (!docSnap.exists()) {
+                return null;
+            }
+            FireStoreObject currentObject = docSnap.toObject(FireStoreObject.class);
+            return makeFSObjectFromFileObject(currentObject);
+
+        } catch (InterruptedException ex) {
+            Thread.currentThread().interrupt();
```

Comment:
That is the best practice for the InterruptedException. The call resets the thread's interrupted flag. See [https://stackoverflow.com/questions/3976344/handling-interruptedexception-in-java](url)

Although it is not the best practice to throw a different exception. Since we're not using interrupted in the server, I think this is OK. But check my logic!